keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,774 @@
+1. `bucketClassName`: (optional) name of the `BucketClass` used to provision this request. If omitted, a default bucket class matching the protocol is used. If the bucket class does not support the requested protocol, an error is logged and the request is retried. A `BucketClass` is required for both greenfield and brownfield requests since BCs support a list of allowed namespaces.
+1. `bucketInstanceName`: (optional) name of the cluster-wide `Bucket` instance. If blank, then COSI assumes this is a greenfield request an will fill in the name during the binding step. If defined by the user, then this names the `Bucket` instance created by the admin. There is no automation available to make this name known to the user.
+1. `bucketAvailable`: if true the bucket has been provisioned. If false then the bucket has not been provisioned and is unable to be accessed.
+
+#### Bucket
+
+A cluster-scoped custom resource representing the abstraction of a single backend bucket. A `Bucket` instance stores enough identifying information so that drivers can accurately target the backend object store (e.g. needed during a deletion process).  Many of the associated bucket class fields are copied to the `Bucket`. Additionally, endpoint data returned by the driver is copied to the `Bucket` by the sidecar.
+
+For greenfield, COSI creates the `Bucket` based on values in the `BucketRequest` and `BucketClass`. For brownfield, an admin manually creates the `Bucket` and COSI copies bucket class fields, populates fields returned by the provisioner, and binds the `Bucket` to the `BucketAccess`.
+
+Since there is a 1:1 mapping between a BR and a `Bucket`, when multiple BRs request access to the same backend bucket, each separate `Bucket` points to the same backend bucket.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: Bucket
+metadata:
+  name: [1]
+  labels: [2]
+    cosi.io/provisioner:
+    cosi.io/bucket-prefix:
+  finalizers:
+    - cosi.io/finalizer [3]
+spec:
+  provisioner: [4]
+  retentionPolicy: [5]
+  anonymousAccessMode: [6]
+  bucketClassName: [7]
+  bucketRequest: [8]
+    name:
+    namespace:
+  allowedNamespaces: [9]
+    - name:
+  protocol: [10]
+    protocolSignature: ""
+    azureBlob:
+      containerName:
+      storageAccount:
+    s3:
+      version:
+      endpoint:
+      bucketName:
+      region:
+      signatureVersion:
+    gcs:
+      bucketName:
+      privateKeyName:
+      projectId:
+      serviceAccount:
+    - name:
+  parameters: [11]
+status:
+  bucketAvailable: [12]
+```
+
+1. `name`: When created by COSI, the `Bucket` name is generated in this format: _"<BR.namespace*>-<BR.name*>-<uuid>"_. If an admin creates a `Bucket`, as is necessary for brownfield access, they can use any name. The uuid is unique within a cluster. `*` the first 10 characters of the namespace and the name are used.
```

Comment:
@jeffvance 10 first characters?? this is way too little. Think about "openshift-" which is already 10 characters long....
Name limit is 253 and uuid is 36. we can limit to 100 first characters and still be on the safe side, no?

Reply:
uuid = 36. `metadata.name` and ns fields are 256 each. So we have 256-36-2 (2 hyphens) avail. So looks like we can use up to 109 chars from ns and from name.